Sources/ShopifyAcceleratedCheckouts/Wallets/ApplePay/ApplePayViewController.swift

```diff
@@ -0,0 +1,266 @@
+            case let .cart(id):
+                guard let cart = try await storefront.cart(by: .init(id)) else {
+                    throw ShopifyAcceleratedCheckouts.Error.invariant(expected: "cart")
+                }
+                return cart
+            case let .variant(id, quantity):
+                let items: [StorefrontAPI.Types.ID] = Array(repeating: .init(id), count: quantity)
+                return try await storefront.cartCreate(with: items)
+            case .invariant:
+                throw ShopifyAcceleratedCheckouts.Error.invariant(expected: "checkoutIdentifier")
+            }
+        } catch let error as StorefrontAPI.Errors {
+            switch error {
+            case let .userError(userErrors, cart):
+                guard let cart else { throw error }
+                let action = ErrorHandler.map(errors: userErrors, shippingCountry: nil, cart: cart)
+                try await handleErrorAction(action: action, cart: cart)
+                return cart
+
+            case let .warning(type, cart):
+                guard let cart else { throw error }
+                let action = ErrorHandler.map(warningType: type, cart: cart)
+                try await handleErrorAction(action: action, cart: cart)
+                return cart
+
+            default:
+                await authorizationDelegate.transition(to: .unexpectedError(error: error))
+                throw error
+            }
+        } catch {
+            await authorizationDelegate.transition(to: .terminalError(error: error))
+            await onFail?()
+            throw error
+        }
+    }
+
+    private func handleErrorAction(
+        /// showError action is not handled uniqely and will present the apple pay sheet with errors
+        action: ErrorHandler.PaymentSheetAction,
+        cart: StorefrontAPI.Types.Cart
+    ) async throws {
+        if case let .interrupt(reason, _) = action {
+            try authorizationDelegate.setCart(to: cart)
+            await authorizationDelegate.transition(to: .interrupt(reason: reason))
+        }
+    }
+
+    func present(url: URL) async throws {
+        let topViewController = await MainActor.run { authorizationDelegate.getTopViewController() }
+
+        guard let topViewController else {
+            throw ShopifyAcceleratedCheckouts.Error.invariant(expected: "topViewController")
+        }
+
+        _ = await MainActor.run {
+            self.checkoutViewController = ShopifyCheckoutSheetKit.present(
+                checkout: url,
+                from: topViewController,
+                delegate: self
+            )
+        }
+    }
+}
+
+@available(iOS 17.0, *)
+extension ApplePayViewController: @preconcurrency CheckoutDelegate {
+    @MainActor func checkoutDidComplete(event _: ShopifyCheckoutSheetKit.CheckoutCompletedEvent) {
+        onComplete?()
+    }
+
+    @MainActor func checkoutDidFail(error _: ShopifyCheckoutSheetKit.CheckoutError) {
+        onFail?()
```

Comment:
Presumably the error should be passed here

Reply:
Same as [here](https://github.com/Shopify/checkout-sheet-kit-swift/pull/327#discussion_r2198211951).